File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.01_02.ClassBox/Box.php
<?php

class Box
{
    /**
     * @var float
     */
    private $length;
    /**
     * @var float
     */
    private $width;
    /**
     * @var float
     */
    private $height;

    /**
     * Box constructor.
     * @param float $length
     * @param float $width
     * @param float $height
     * @throws Exception
     */
    public function __construct(float $length, float $width, float $height)
    {
        $this->setLength($length);
        $this->setWidth($width);
        $this->setHeight($height);
    }

    /**
     * @return float
     */
    public function getLength(): float
    {
        return $this->length;
    }

    /**
     * @param float $length
     * @throws Exception
     */
    private function setLength(float $length): void
    {
        $this->validateInput($length, "Length");
        $this->length = $length;
    }

    /**
     * @return float
     */
    public function getWidth(): float
    {
        return $this->width;
    }

    /**
     * @param float $width
     * @throws Exception
     */
    private function setWidth(float $width): void
    {
        $this->validateInput($width, "Width");
        $this->width = $width;
    }

    /**
     * @return float
     */
    public function getHeight(): float
    {
        return $this->height;
    }

    /**
     * @param float $height
     * @throws Exception
     */
    private function setHeight(float $height): void
    {
        $this->validateInput($height, "Height");
        $this->height = $height;
    }

    private function surfaceArea() : float
    {
        return  2 * $this->getLength() * $this->getWidth() + 2 * $this->getLength() * $this->getHeight() + 2 * $this->getWidth() * $this->getHeight();
    }
    private function lateralSurfaceArea() : float
    {
        return 2 * $this->getLength() * $this->getHeight() + 2 * $this->width * $this->getHeight();
    }
    private function volume() : float
    {
        return $this->getLength() * $this->getWidth() *$this->getHeight();
    }

    /**
     * @param float $value
     * @param string $param
     * @throws Exception
     */
    private function validateInput(float $value, string $param) : void
    {
        if ($value <= 0) {
            throw new Exception("{$param} cannot be zero or negative.");
        }
    }
    public function __toString()
    {
        return "Surface Area - " . number_format($this->surfaceArea(), 2, ".", "").PHP_EOL
            ."Lateral Surface Area - " . number_format($this->lateralSurfaceArea(), 2, ".", "") . PHP_EOL
            . "Volume - " .number_format($this->volume(), 2, ".", "");

    }

}

File: Courses/C# OOP/08. Encapsulation Exercise/08. Encapsulation Exercise/02.ClassBoxDataValidation/Box.cs
using System;

namespace ClassBox
{
    public class Box
    {
        private decimal length;
        private decimal width;
        private decimal height;

        public Box(decimal length, decimal width, decimal height)
        {
            Length = length;
            Width = width;
            Height = height;
        }

        public decimal Length 
        { 
            get => length;
            private set
            {
                ValidateSide("Length", value);
                //if (value <= 0)
                //{
                //    throw new ArgumentException("Length cannot be zero or negative.");
                //}
                this.length = value;
            }
        }
        public decimal Width 
        {
            get => width;
            private set
            {
                ValidateSide("Width", value);
                //if (value <= 0)
                //{
                //    throw new ArgumentException("Width cannot be zero or negative.");
                //}
                this.width = value;
            }
        }
        public decimal Height 
        {
            get => height;
            private set
            {
                ValidateSide("Height", value);
                //if (value <= 0)
                //{
                //    throw new ArgumentException("Height cannot be zero or negative.");
                //}
                this.height = value;
            }
        }

        private void ValidateSide(string sideName, decimal value)
        {
            if (value <= 0)
            {
                throw new ArgumentException($"{sideName} cannot be zero or negative.");
                //Console.WriteLine($"{sideName} cannot be zero or negative.");
            }
        }
        public decimal SurfaceArea()
        {
            return 2* Length * Width + 2 * Length * Height + 2 * Width * Height;
        }
        public decimal LateralSurfaceArea()
        {
            return 2 * Length * Height + 2 * Width * Height;
        }
        public decimal Volume()
        {
            return Length * Width * Height;
        }
    }
}
